contracts/bridge/contracts/BasicInboundChannel.sol

```diff
@@ -35,57 +32,63 @@ contract BasicInboundChannel {
     function submit(
         Message[] memory messages,
         bytes memory beefyMMRLeaf,
+        bytes memory blockHeader,
         uint256 beefyMMRLeafIndex,
         uint256 beefyMMRLeafCount,
         bytes32[] memory peaks,
         bytes32[] memory siblings 
     ) public {
+        bytes32 beefyMMRLeafHash = keccak256(beefyMMRLeaf);
         require(
             lightClientBridge.verifyBeefyMerkleLeaf(
-                beefyMMRLeaf,
+                beefyMMRLeafHash,
                 beefyMMRLeafIndex,
                 beefyMMRLeafCount,
                 peaks,
                 siblings
             ),
-            "Invalid proof"
+            "Channel: Invalid proof"
         );
-        verifyMessages(messages, beefyMMRLeaf);
+        verifyMessages(messages, beefyMMRLeaf, blockHeader);
         processMessages(messages);
     }
 
-    // struct BlockHeader {
-    //     parentHash bytes32;
-    //     number uint64;
-    //     stateRoot bytes32;
-    //     extrinsicsRoot bytes32;
-    //     digest bytes;
-    //     messagesRoot bytes32;
-    // }
     //TODO: verifyMessages should accept all needed proofs
-    function verifyMessages(Message[] memory messages, bytes memory /*beefyMMRLeaf*/)
+    function verifyMessages(Message[] memory messages, bytes memory beefyMMRLeaf, bytes memory blockHeader)
         internal
         view
         returns (bool success)
     {
-
-        // Scale.decodeBeefyMMRLeaf(beefyMMRLeaf)
-        // Scale.decodeBlockHeader(beefyMMRLeaf.BlockHeader)
-        // require(
-        //     blockHeader.BlockNumber <= lightClientBridge.getFinalizedBlockNumber(),
-        //     "block not finalized"
-        // )
+        // struct BeefyMMRLeaf {
+        //     uint32 parentNumber;
+        //     bytes32 parentHash;
+        //     bytes32 parachainHeadsRoot;
+        //     uint64 nextAuthoritySetId;
+        //     uint32 nextAuthoritySetLen;
+        //     bytes32 nextAuthoritySetRoot;
+        // }
+        // struct MMRLeaf {
+        //     bytes32 parentHash;
+        // }
+        bytes32 blockHash = abi.decode(beefyMMRLeaf, (bytes32));
+        require(blockHash == keccak256(blockHeader), "Channel: invalid block header");
+        uint32 blockNumber = ScaleHeader.decodeBlockNumberFromBlockHeader(blockHeader);
+        require(
+            blockNumber <= lightClientBridge.getFinalizedBlockNumber(),
+            "Channel: block not finalized"
+        );
+        bytes32 messagesRoot = ScaleHeader.decodeMessagesRootFromBlockHeader(blockHeader);
```

Comment:
We could put message root directly in the mmr leaf, this version will be changes later to final design.

LGTM, let's first merge this version for now.

Reply:
https://github.com/darwinia-network/darwinia-bridge-sol/issues/43
I will do this in next pr.

Reply:
I complete this in this [commit](https://github.com/darwinia-network/darwinia-bridge-sol/pull/41/commits/5b642519329025b53caee0276b19d3b96b04bbe3)
@HackFisher

Reply:
Don't need another pr now.